index.bs

```diff
@@ -277,22 +277,18 @@ interface mixin UnidirectionalStreamsTransport {
    run the following steps:
      1. Let |transport| be the `UnidirectionalStreamsTransport` on which
         `createUnidirectionalStream` is invoked.
+     1. Let |session| be |transport|'s [=[[Session]]=].
      1. If |transport|'s {{WebTransport/state}} is `"closed"` or `"failed"`,
         immediately return a new [=rejected=] promise with a newly created
         {{InvalidStateError}} and abort these steps.
      1. Let |p| be a new promise.
      1. Return |p| and continue the remaining steps [=in parallel=].
-     1. [=SendStream/Create=] a {{SendStream}} with |transport| and |p| when all
-        of the following conditions are met:
-         1. The |transport|'s {{WebTransport/state}} is `"connected"`.
-         1. Stream creation flow control is not being violated by exceeding the
-            max stream limit set by the remote endpoint, as specified in
-            [[!QUIC]].
-         1. |p| has not been [=settled=].
-     1. Queue a task to [=reject=] |p| with a newly created {{InvalidStateError}}
-        when all of the following conditions are met:
-         1. The |transport|'s state is `"closed"` or `"failed"`.
-         1. |p| has not been [=settled=].
+     1. Let |interanlStream| be the result of [=creating an outgoing stream=] with
+        |session|. If this operation fails, [=reject=] |p| with a {{TypeError}} and
```

Comment:
When we already have [too many streams](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-34#section-4.6), for example.

By the way, we have two ways to represent an error - a rejected promise and a promise resolved with an errored stream - which is generally bad. Is it possible to have `createUnidirectionalStream` return just a `SendStream`, rather than a `Promise<SendStream>`? Sorry if this has already been discussed (I'm happy to file a new issue if it's better).

@ricea @vasilvv

Reply:
Ergonomically I think it would be pretty good, although it might be non-obvious that you need to wait for `writer.ready` to find out when you can actually send data. My main concern is that a common pattern might be:
```js
const stream = wt.createUnidirectionalStream();
const writer = stream.getWriter();
writer.write(message);
writer.close();
```
It's not obvious from looking at the code that error-handling is missing. It will appear to work during testing, so errors might only occur when deployed. And unless the page hooks the "unhandledrejection" event then only evidence that anything has gone wrong will be a console log.

By the way, if we have two many streams I think the behaviour should be to wait for the server to provide more, not to error.

Reply:
> Ergonomically I think it would be pretty good, although it might be non-obvious that you need to wait for writer.ready to find out when you can actually send data.

Good point. But in your example the write function raises an exception, right?

> By the way, if we have two many streams I think the behaviour should be to wait for the server to provide more, not to error.

Given the following descriptions, I believe we should raise an error when we reach the max stream.

>   An endpoint limits the cumulative number of incoming streams a peer
>   can open.  Only streams with a stream ID less than (max_stream * 4 +
>   initial_stream_id_for_type) can be opened
> [4.6 Controlling Concurrency](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-34#section-4.6)

> A QUIC endpoint MUST NOT reuse a stream ID within a connection.
> [2.1 Stream Types and Identifiers](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-34#section-2.1)

Reply:
> Good point. But in your example the write function raises an exception, right?

@yutakahirano [write](https://streams.spec.whatwg.org/#default-writer-class-definition) returns a promise, which means there is no way for it to throw.

Reply:
Ah yes that's right, thank you.

Reply:
Section 4.6 also says:

>    An endpoint that is unable to open a new stream due to the peer's
>    limits SHOULD send a STREAMS_BLOCKED frame (Section 19.14).  This
>    signal is considered useful for debugging.  An endpoint MUST NOT wait
>    to receive this signal before advertising additional credit, since
>    doing so will mean that the peer will be blocked for at least an
>    entire round trip, and potentially indefinitely if the peer chooses
>    not to send STREAMS_BLOCKED frames.

To me, "the peer will be blocked for at least an entire round trip" implies that the peer waits for a MAX_STREAMS frame rather that erroring.

Also, if `createUnidirectionalStream()` errored when `max_stream` was reached, how would JavaScript know when the limit was increased? It would have to call the method repeatedly until it succeeded, which would be very inefficient.

Reply:
I see, I fixed the logic.